File: sdk/ruby/test/wire/a2a_test.rb
# frozen_string_literal: true

require_relative "wiremock_test_case"

class A2aWireTest < WireMockTestCase
  def setup
    super

    @client = AgentMailer::Client.new(
      api_key: "<token>",
      base_url: WIREMOCK_BASE_URL
    )
  end

  def test_a2a_send_task_with_wiremock
    test_id = "a2a.send_task.0"

    @client.a2a.send_task(
      handle: "handle",
      jsonrpc: "2.0",
      method_: "method",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "a2a.send_task.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "POST",
      url_path: "/a2a/handle",
      query_params: nil,
      expected: 1
    )
  end

  def test_a2a_get_agent_card_with_wiremock
    test_id = "a2a.get_agent_card.0"

    @client.a2a.get_agent_card(
      handle: "handle",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "a2a.get_agent_card.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "GET",
      url_path: "/a2a/handle/.well-known/agent-card.json",
      query_params: nil,
      expected: 1
    )
  end

  def test_a2a_update_task_with_wiremock
    test_id = "a2a.update_task.0"

    @client.a2a.update_task(
      task_id: "taskId",
      state: "TASK_STATE_SUBMITTED",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "a2a.update_task.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "POST",
      url_path: "/v1/a2a/tasks/taskId",
      query_params: nil,
      expected: 1
    )
  end

  def test_a2a_update_identity_with_wiremock
    test_id = "a2a.update_identity.0"

    @client.a2a.update_identity(request_options: {
      additional_headers: {
        "X-Test-Id" => "a2a.update_identity.0"
      }
    })

    verify_request_count(
      test_id: test_id,
      method: "PATCH",
      url_path: "/v1/a2a/identity",
      query_params: nil,
      expected: 1
    )
  end
end

File: sdk/rust/tests/a2a_test.rs
use agentmailer::prelude::*;

mod wire_test_utils;

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_a2a_send_task_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentMailer::new(config).expect("Failed to build client");

    let result = client
        .a2a
        .send_task(
            &"handle".to_string(),
            &A2AJsonRpcRequest {
                jsonrpc: A2AJsonRpcRequestJsonrpc::Two0,
                method: "method".to_string(),
                id: None,
                params: None,
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("POST", "/a2a/handle", None, 1)
        .await
        .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_a2a_get_agent_card_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentMailer::new(config).expect("Failed to build client");

    let result = client.a2a.get_agent_card(&"handle".to_string(), None).await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count(
        "GET",
        "/a2a/handle/.well-known/agent-card.json",
        None,
        1,
    )
    .await
    .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_a2a_update_task_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentMailer::new(config).expect("Failed to build client");

    let result = client
        .a2a
        .update_task(
            &"taskId".to_string(),
            &A2ATaskUpdate {
                state: A2ATaskUpdateState::TaskStateSubmitted,
                message: None,
                artifacts: None,
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("POST", "/v1/a2a/tasks/taskId", None, 1)
        .await
        .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_a2a_update_identity_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentMailer::new(config).expect("Failed to build client");

    let result = client
        .a2a
        .update_identity(
            &A2AIdentitySettings {
                ..Default::default()
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("PATCH", "/v1/a2a/identity", None, 1)
        .await
        .unwrap();
}
